File: memento/golang/waiter.go
package main

// Waiter represents another Originator
type Waiter struct {
	ActiveTable int
}

// WaiterMemento stores the Waiter state
type WaiterMemento struct {
	activeTable int
}

func (w *Waiter) Save() *WaiterMemento {
	return &WaiterMemento{activeTable: w.ActiveTable}
}

func (w *Waiter) Restore(m *WaiterMemento) {
	w.ActiveTable = m.activeTable
}

File: memento/rust/memento/src/waiter.rs
#[derive(Debug)]
pub struct Waiter {
    pub status: String, // e.g., "Idle", "Serving", "Cleaning"
}

pub struct WaiterMemento {
    state: String,
}

impl Waiter {
    pub fn new() -> Self {
        Self { status: "Idle".to_string() }
    }

    pub fn save(&self) -> WaiterMemento {
        WaiterMemento { state: self.status.clone() }
    }

    pub fn restore(&mut self, memento: &WaiterMemento) {
        self.status = memento.state.clone();
    }
}
